src/Chapter3.hs

```diff
@@ -480,6 +546,81 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city in total.
 -}
 
+data Person 
+    = Person 
+    deriving (Show)
+
+data House = House 
+    { houseResidents :: [Person]
+    } deriving (Show)
+
+data CityBuilding 
+    = Church 
+    | Library 
+    deriving (Show)
+
+data Castle = Castle
+    { castleName    :: String,
+      cityWalls     :: Int 
+    } deriving (Show)
+
+
+data MagicalCity = MagicalCity
+    { cityBuilding    :: CityBuilding,
+      houses          :: [House],
+      cityCastle      :: Maybe Castle
+    } deriving (Show)
+
+isChurch :: CityBuilding -> Bool
+isChurch (Church) = True
+isChurch _ = False 
+
+hasChurch :: MagicalCity -> Bool
+hasChurch mC = isChurch (cityBuilding mC)
+
+hasCastle :: MagicalCity -> Bool
+hasCastle mC = not (isNothing (cityCastle mC))
+
+wallCount :: Maybe Castle -> Int
+wallCount c
+    | not (isNothing c) = cityWalls (fromJust c)
+    | otherwise = 0
+
+cityWallsCount :: MagicalCity -> Int
+cityWallsCount mC = wallCount (cityCastle mC)
+
+cityResidentsCount :: MagicalCity -> Int
+cityResidentsCount mC = sum (map (\x -> length (houseResidents x)) (houses mC))
+
+testCity :: MagicalCity
+testCity = MagicalCity 
+    { cityBuilding = Church, 
+    houses = [ House { houseResidents = [Person, Person] }, House { houseResidents = [Person, Person] } ],
+    cityCastle = Just (Castle {castleName = "Magical Castle", cityWalls = 5}) 
+    }
+-- testCity = MagicalCity { cityBuilding = Church, houses = [ House { houseResidents = [Person, Person, Person, Person, Person, Person] }, House { houseResidents = [Person, Person, Person, Person, Person, Person] } ], cityCastle = Just (Castle {castleName = "Magical Castle", cityWalls = 5}) }
+-- testCity = MagicalCity { cityBuilding = Church, houses = [House [Person]], cityCastle = Nothing }
+
+buildCastle :: MagicalCity -> String -> MagicalCity
+buildCastle mC castleN = mC { cityCastle = newCastle mC castleN}
+    where
+      newCastle :: MagicalCity -> String -> Maybe Castle
+      newCastle mC castleN = if hasCastle mC then Just ((fromJust (cityCastle mC)) { castleName = castleN}) else Just (Castle {castleName = castleN, cityWalls = 0}) 
+
+buildHouse :: MagicalCity -> MagicalCity
+buildHouse mC =
+    let peopleNum = if hasCastle mC then if hasChurch mC then 4 else 3 else if hasChurch mC then 2 else 1
+    in mC {houses = House (replicate peopleNum Person) : houses mC}
```

Comment:
Interesting rules 🤔 
Note, that an alternative design could be having a number as an input 🙂